### DISCUSSION TITLE
Adding Different Resource Views?

### DISCUSSION THREAD

**Discussion by Joebjammin** [ckan/ckan#6643] [Q&A] [answered]
Hi,
I would like to replace recline views and insert plotly charts etc. Is this possible? If so, does anyone have any guidance.
Much obliged.
...Joe

**Comment by amercader** [comment, marked as answer]
It is certainly possible. The available views depend on which view plugins you have loaded in your ini file. Aside from the [default core ones](http://docs.ckan.org/en/latest/maintaining/data-viewer.html#available-view-plugins) you can install and load any existing third party ones, like:

* https://github.com/keitaroinc/ckanext-visualize
* https://github.com/datopian/ckanext-dataexplorer-react

And if you want to write your own you can use the [`IResourceView`](http://docs.ckan.org/en/latest/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IResourceView) interface. The examples above or one of the [simpler ones](https://github.com/ckan/ckan/tree/master/ckanext/textview) included in core should be useful to get started.

**Reply by Joebjammin** [reply]
Thank you very much. Exactly what I was looking for and hoping to hear.

**Comment by pdekraker-epa** [comment]
Another extension that might be useful is [ckanext-plotly](https://github.com/USEPA/ckanext-plotly)

**Reply by Joebjammin** [reply]
Thanks. This will be very helpful.

**Reply by BENETNATH** [reply]
> Another extension that might be useful is [ckanext-plotly](https://github.com/USEPA/ckanext-plotly)

Did you manage to make it working ?

**Comment by Mohab25** [comment]
@amercader is it possible to add multiple custom views from one extension, IResourceView info() method returns a single dict holding  configuration options for one custom view only ?

**Reply by amercader** [reply]
The interface was designed with a single plugin view in mind. In theory we could refactor the plugin hooks in CKAN core to accept list of dicts in addition to single dicts but I'm not sure how this would work with the other interface methods, so not sure is worth exploring.

**Reply by pdekraker-epa** [reply]
A single extension can however implement multiple plugins, it just means each plugin / resource view will need to be listed in the ini file. [ckanext-plotly](https://github.com/USEPA/ckanext-plotly) is an example of this and there are others. If you look at its [plugin.py](https://github.com/USEPA/ckanext-plotly/blob/master/ckanext/plotly/plugin.py) you will see two separate plugin classes one for each view.

**Reply by Mohab25** [reply]
Thank you both @amercader and @pdekraker-epa , sure will be checking ckanext-plotly.

**Reply by BENETNATH** [reply]
Did you succeed in having plotly working in CKAN ? I'm stuck on it

**Reply by Mohab25** [reply]
just to complete this for future readers, you need to make a new class that extends `plugins.SingletonPlugin` and implements [`IResourceView`](https://docs.ckan.org/en/2.9/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IResourceView),  register that class in `entry_points` found in `setup.py`,  run `python setup.py develop` again, and add the plugin to the `ckan.ini` , 
if you need your view to be enabled by default with specific files extensions (e.g. `geojson`), an extra step is to add your view to  `ckan.views.default_views` in the `ckan.ini`, notice the view name might be different from the plugin name
